File: UVA judge/331.py
from sys import stdin, stdout, exit
from collections import deque
from math import inf

def main():
    cin = Scanner()
    array = None
    sortList = None


    def searchPair():
        answer = 0
        if array == sortList:
            return 1

        for i in range(n - 1):
            if array[i] > array[i + 1]:
                array[i], array[i + 1] = array[i + 1], array[i]
                answer += searchPair()
                array[i], array[i + 1] = array[i + 1], array[i]

        return answer

    c = 1
    line = cin.next_array(int)
    while line != [0]:
        n = line[0]
        array = line[1:]
        sortList = sorted(array)

        answer = 0 if array == sortList else searchPair()

        print(f"There are {answer} swap maps for input data set {c}.")
        c += 1

        line = cin.next_array(int)


class Scanner:

    def next_int(self):
        return int(self.next_line())

    def next_float(self):
        return float(self.next_line())

    def next_line(self):
        return stdin.readline().strip()

    def next_array(self, fun = str):
        line = self.next_line().split(' ')
        return line if fun is str else list(map(fun, line))


if __name__ == "__main__":
    main()

File: UVA judge/331.cpp
#include <bits/stdc++.h>
#define all(x) x.begin(),x.end()
#define ff first
#define ss second
#define mt make_tuple
#define eb emplace_back

using namespace std;

typedef long long ll;
typedef vector<ll> vi;
typedef pair<ll,ll> pii;
typedef vector<pii> vii;
const ll INF = numeric_limits<ll>::max();

ll n, t = 1, answer;
vi sequence;

bool isSorted() {
  for (size_t i = 0; i < n - 1; i++) {
    if (sequence[i] > sequence[i + 1])
      return false;
  }
  return true;
}

void seePair() {
  if (isSorted()) {
    answer++;
    return;
  }

  for (size_t i = 0; i < n - 1; i++) {
    if (sequence[i] > sequence[i + 1]) {
      swap(sequence[i], sequence[i + 1]);
      seePair();
      swap(sequence[i], sequence[i + 1]);
    }
  }
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(NULL);

  while (cin >> n, n) {

    answer = 0;
    sequence = vi(n);
    for (size_t i = 0; i < n; i++) {
      cin >> sequence[i];
    }
    if (!isSorted())
      seePair();
    cout << "There are " << answer << " swap maps for input data set " << t++ << ".\n";
  }
  return 0;
}
